## Part 1: Explore the Data

Import the data and use Pandas to learn more about the dataset.

In [39]:
import pandas as pd

df = pd.read_csv('Resources/client_dataset.csv')
df.head()

first      last                    job               phone  \
0     Donald   Harding           Immunologist  793-904-7725x39308   
1    Tiffany     Myers        Music therapist    201.442.4543x942   
2    Shannon    Watson           Immunologist   687.737.9424x8503   
3     Nathan     Baker  Accounting technician    827-788-8123x012   
4  Christina  Schwartz           Chiropractor        265-829-3643   

                                 email  client_id  order_id  order_date  \
0     harding.donald.7185@sullivan.com      58515   8953482  2023-04-28   
1    myers.t.6537@ferguson-johnson.net      37609   8069089  2023-05-19   
2                swatson8146@payne.net      57113   1902144  2023-01-29   
3               bakernathan@benson.com      46554   9031802  2023-04-25   
4  christinaschwartz9252@mcconnell.com      92089   1322274  2023-05-28   

   order_week  order_year         item_id     category         subcategory  \
0          17        2023  EUD29711-63-6U        decor            wall art   
1          20        2023  XDA18116-89-4A  consumables                pens   
2           4        2023  ABE59463-05-7E     software  project management   
3          17        2023  ZMM00836-65-0C  consumables                pens   
4          21        2023  BZX55559-12-3X  consumables                misc   

   unit_price  unit_cost  unit_weight  qty  line_number  
0     1096.80     762.71         7.50  105            1  
1       24.95      15.09         1.49   21            0  
2       13.52       7.86         1.68   39            6  
3       36.42      24.85         1.23   29            3  
4      195.10     108.17        46.43   20            1

In [40]:
# View the column names in the data
df.columns

Index(['first', 'last', 'job', 'phone', 'email', 'client_id', 'order_id',
       'order_date', 'order_week', 'order_year', 'item_id', 'category',
       'subcategory', 'unit_price', 'unit_cost', 'unit_weight', 'qty',
       'line_number'],
      dtype='object')

In [41]:
# Use the describe function to gather some basic statistics
df.describe()

client_id      order_id    order_week    order_year    unit_price  \
count  54639.000000  5.463900e+04  54639.000000  54639.000000  54639.000000   
mean   54837.869416  5.470190e+06     11.359139   2022.993064    136.267207   
std    25487.438231  2.599807e+06      7.023499      0.082997    183.873135   
min    10033.000000  1.000886e+06      1.000000   2022.000000      0.010000   
25%    33593.000000  3.196372e+06      6.000000   2023.000000     20.800000   
50%    53305.000000  5.496966e+06     11.000000   2023.000000     68.310000   
75%    78498.000000  7.733869e+06     17.000000   2023.000000    173.160000   
max    99984.000000  9.998480e+06     52.000000   2023.000000   1396.230000   

          unit_cost   unit_weight           qty   line_number  
count  54639.000000  54639.000000  5.463900e+04  54639.000000  
mean      99.446073      5.004116  5.702646e+02      2.979667  
std      133.164267      5.326599  1.879552e+04      2.436320  
min        0.010000      0.000000  0.000000e+00      0.000000  
25%       14.840000      1.450000  3.200000e+01      1.000000  
50%       49.890000      3.240000  6.800000e+01      3.000000  
75%      125.570000      6.890000  1.700000e+02      5.000000  
max      846.270000     46.430000  3.958244e+06      9.000000

In [232]:
# Use this space to do any additional research
# and familiarize yourself with the data.
renamed_df = df.rename(columns={'first': 'First Name', 
                                'last': 'Last Name', 
                                'job': 'Job', 
                                'phone': 'Phone', 
                                'email': 'Email', 
                                'client_id' : 'Client ID', 
                                'order_id' : 'Order ID', 
                                'order_date' : 'Order Date', 
                                'order_week' : 'Order Week',
                                'order_year' : 'Order Year',
                                'item_id' : 'Item ID',
                                'category' : 'Category',
                                'subcategory' : 'Subcategory',
                                'unit_price' : 'Unit Price',
                                'unit_cost' : 'Unit Cost',
                                'unit_weight' : 'Unit Weight',
                                'qty' : 'Quantity',
                                'line_number' : 'Line Number'})
renamed_df.count()

First Name     54639
Last Name      54639
Job            54639
Phone          54639
Email          54639
Client ID      54639
Order ID       54639
Order Date     54639
Order Week     54639
Order Year     54639
Item ID        54639
Category       54639
Subcategory    54639
Unit Price     54639
Unit Cost      54639
Unit Weight    54639
Quantity       54639
Line Number    54639
dtype: int64

In [233]:
# What three item categories had the most entries?

item_entries = renamed_df["Category"].value_counts()
top_3_categories = item_entries.head(3)
top_3_category_names = top_3_categories.index
print(f'The top three categories with the most entries are {top_3_category_names.tolist()}.')


The top three categories with the most entries are ['consumables', 'furniture', 'software'].


In [234]:
# For the category with the most entries, which subcategory had the most entries?
renamed_df.set_index(renamed_df['Category'], inplace= True)
consumables_df = renamed_df.sort_index()

consumable_entries = consumables_df["Subcategory"].value_counts()
top_consumable_subcategory = consumable_entries.head(1)
top_consumable_subcategory_name = top_consumable_subcategory.index
print(f'The subcategory with the most entries, in the category with the most entries (consumables), is {top_consumable_subcategory_name.to_list()}.')

The subcategory with the most entries, in the category with the most entries (consumables), is ['bathroom supplies'].


In [235]:
# Which five clients had the most entries in the data?
client_entries = renamed_df['Client ID'].value_counts()
top_5_clients = client_entries.head()
top_5_clients_names = top_5_clients.index


#print(f'The top five clients with most entries are {top_5_clients_names.to_list()}.')

The top five clients with most entries are [33615, 66037, 46820, 38378, 24741].


In [236]:
# Store the client ids of those top 5 clients in a list.


In [237]:
# How many total units (the qty column) did the client with the most entries order order?


## Part 2: Transform the Data
Do we know that this client spent the more money than client 66037? If not, how would we find out? Transform the data using the steps below to prepare it for analysis.

In [238]:
# Create a column that calculates the subtotal for each line using the unit_price and the qty


In [239]:
# Create a column for shipping price.
# Assume a shipping price of $7 per pound for orders over 50 pounds and $10 per pound for items 50 pounds or under.



In [240]:
# Create a column for the total price using the subtotal and the shipping price along with a sales tax of 9.25%


In [241]:
# Create a column for the cost of each line using unit cost, qty, and
# shipping price (assume the shipping cost is exactly what is charged to the client).



In [242]:
# Create a column for the profit of each line using line cost and line price



## Part 3: Confirm your work
You have email receipts showing that the total prices for 3 orders. Confirm that your calculations match the receipts. Remember, each order has multiple lines.

Order ID 2742071 had a total price of \$152,811.89

Order ID 2173913 had a total price of \$162,388.71

Order ID 6128929 had a total price of \$923,441.25


In [243]:
# Check your work using the totals above


## Part 4: Summarize and Analyze
Use the new columns with confirmed values to find the following information.

In [244]:
# How much did each of the top 5 clients by quantity spend? Check your work from Part 1 for client ids.



In [245]:
# Create a summary DataFrame showing the totals for the for the top 5 clients with the following information:
# total units purchased, total shipping price, total revenue, and total profit. 


In [246]:
# Format the data and rename the columns to names suitable for presentation.

# Define the money columns. 


# Define a function that converts a dollar amount to millions.


# Apply the currency_format_millions function to only the money columns. 


# Rename the columns to reflect the change in the money format. 




In [247]:
# Sort the updated data by "Total Profit (millions)" form highest to lowest and assign the sort to a new DatFrame.
